### ISSUE TITLE
fade_brightness() method throws many different LookupError exceptions

### ISSUE THREAD

**Issue by lazzarello** [Crozzers/screen_brightness_control#18]
The `fade_brightness()` method works > 80% of the time but has many different exception messages when `filter_monitors` throws a LookupError. They all seem to be caused by unexpected output from `ddcutil`. Here's three examples and my guesses as to what could be happening. The monitor cable is attached all throughout these exceptions and I have an image on the screen.

The first few API calls worked, then one that seems to have worked but the (possibly) cached value does not match the brightness value specified. Then it looks like the display is gone.

```
>>> sbc.set_brightness(60, display=1)
>>> sbc.fade_brightness(25, interval=1, increment=5, display=1)
[25]
>>> sbc.fade_brightness(60, interval=1, increment=5, display=1)
[60]
>>> sbc.fade_brightness(25, interval=0.75, increment=5, display=1)
[24]
>>> sbc.fade_brightness(60, interval=0.75, increment=5, display=1)
[24] # This did not fade to 60%. Both the output value and the percieved brightness did not change. 
>>> sbc.fade_brightness(60, interval=0.75, increment=5, display=1)
[60] 
>>> sbc.fade_brightness(60, interval=0.25, increment=5, display=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 251, in fade_brightness
    return get_brightness(**kwargs)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 44, in get_brightness
    return __brightness(display=display, method=method, meta_method='get', verbose_error=verbose_error)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 769, in __brightness
    raise ScreenBrightnessError(msg)
screen_brightness_control.helpers.ScreenBrightnessError:
        filter_monitors -> LookupError: no monitors found with name/serial/model/edid/index of 1
```

The next exception seems like the I2C device just...disappears.

```
>>> sbc.fade_brightness(65, display=1)
No monitor detected on I2C bus /dev/i2c-8
Exception in thread Thread-18:
Traceback (most recent call last):
  File "/home/lee/.pyenv/versions/3.9.7/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/home/lee/.pyenv/versions/3.9.7/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 180, in fade
    monitor.set_brightness(val, no_return=True)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 481, in set_brightness
    self.method.set_brightness(value, display=self.index)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/linux.py", line 1150, in set_brightness
    subprocess.check_call(
  File "/home/lee/.pyenv/versions/3.9.7/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['ddcutil', 'setvcp', '10', '21626', '-b', '8', '--sleep-multiplier=0.5']' returned non-zero exit status 1.
[32] 
```

And finally...it looks like the I2C device is back but it seems that the EDID data is passed in hex to the index lookup rather then an int. If it helps this data is different if I export the EDID through the nvidia-settings application. The cached brightness value looks like it updated for a few loops and stopped with the exception.

```
>>> sbc.fade_brightness(65, display=1)
Exception in thread Thread-20:
Traceback (most recent call last):
  File "/home/lee/.pyenv/versions/3.9.7/lib/python3.9/threading.py", line 973, in _bootstrap_inner
    self.run()
  File "/home/lee/.pyenv/versions/3.9.7/lib/python3.9/threading.py", line 910, in run
    self._target(*self._args, **self._kwargs)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 180, in fade
    monitor.set_brightness(val, no_return=True)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 479, in set_brightness
    self.get_info()
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 575, in get_info
    info = filter_monitors(display=identifier[1], method=self.method.__name__)[0]
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 691, in filter_monitors
    raise LookupError(msg)
LookupError: no monitors found with name/serial/model/edid/index of '00ffffffffffff0006102792df0e0d170d170104b53c2278226fb1a7554c9e250c5054000000010101010
10101010101010101010101565e00a0a0a029503020350055502100001a1a1d008051d01c204080350055502100001c000000ff004330324b4733395a463247430a000000fc005468756e64657
2626f6c740a20014e' with method of "DDCUtil"
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 251, in fade_brightness
    return get_brightness(**kwargs)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 44, in get_brightness
    return __brightness(display=display, method=method, meta_method='get', verbose_error=verbose_error)
  File "/home/lee/.local/lib/python3.9/site-packages/screen_brightness_control/__init__.py", line 769, in __brightness
    raise ScreenBrightnessError(msg)
screen_brightness_control.helpers.ScreenBrightnessError:
        filter_monitors -> LookupError: no monitors found with name/serial/model/edid/index of 1

>>> sbc.get_brightness(display=1)
[44] 
```

**Comment by Crozzers** [comment]
Well this is interesting.

The main thing here seems to be weird behaviour with `DDCUtil` and if we could see the stdout and stderr then I think it would explain alot.
I've pushed a commit to an [issue18 branch](https://github.com/Crozzers/screen_brightness_control/tree/issue18) that should print out stdout and stderr upon `CalledProcessError`s so if you could try that and report back the output that would be great.

As for some of the other things you mentioned:
> [...] it seems that the EDID data is passed in hex to the index lookup rather then an int.

This is intentional behaviour. `fade_brightness` works a little bit differently to `set_brightness` and `get_brightness` in that it uses the `Monitor` class to adjust each display's brightness individually. When `Monitor.set_brightness` is called it will try to refresh the information it has about itself to ensure it is up to date. Part of this is finding an data point (EDID, name, serial number or index) that is likely to be unique to the monitor and is a known value. This identifier is then passed to `filter_monitors` to return the current information about that monitor.
In this case, the EDID string was passed as the identifier to `filter_monitors` and a `LookupError` was raised because it could not find any monitors with that EDID string, meaning that `DDCUtil.get_display_info` likely did not report the displays correctly.

Secondly, in your first code example you show this snippet:
```
>>> sbc.fade_brightness(25, interval=0.75, increment=5, display=1)
[24]
```
Which demonstrates the brightness not quite being set to an accurate value. This is a bug I have been aware of for a little while and I think it has something to do with caches not being expired on time.

**Comment by Crozzers** [comment]
After doing a couple of unrelated tests on the library, I actually began to run into the above mentioned issues. Using the [issue18 branch](https://github.com/Crozzers/screen_brightness_control/tree/issue18) I was able to get the following outputs:
```
============STDOUT============
b''
============STDERR============
b'Verification failed for feature 10\n'
```
```
============STDOUT============
b'DDC communication failed for monitor on bus /dev/i2c-7\n'
============STDERR============
b''
```
I'm going to guess that these messages are the result of `fade_brightness` spamming the monitors and this has a chance of causing issues.
There are a few approaches to tackle this:
1. Use `--noverify` when setting brightness with DDCUtil (see [ddcutil performance options](https://www.ddcutil.com/performance_options/))
    * But not always. Perhaps enable this option during `fade_brightness` but turn it off for a final check
2. If the command fails then retry after a short wait
3. If the command fails then increase the `DDCUtil.sleep_multiplier` value
4. A combination of the above

I'll muck about and see if these make a viable solution

**Comment by lazzarello** [comment]
Wooh, thanks. I took a week off open source things but I'm back. I'll rebuild and give it a shot.

**Comment by lazzarello** [comment]
lgtm! Hardware responded as expected. Thanks for this. I find a special kind of satisfaction getting obsolete Apple hardware to work better on non-apple computers.

```
>>> import screen_brightness_control as sbc
>>> print(sbc.fade_brightness(100, start=0, method='ddcutil', verbose_error=True))
[100]
>>> print(sbc.fade_brightness(0, method='ddcutil', verbose_error=True))
[0]
>>> print(sbc.fade_brightness(100, method='ddcutil', verbose_error=True))
[100]
>>> sbc.set_brightness(60, method='ddcutil', verbose_error=True)
```

**Comment by Crozzers** [comment]
Wonderful stuff. The issue18 branch was merged into main and pushed out in v0.14.0.
Thanks for raising the issue
